src/sparql/QueryObjectBuilder.ts

```diff
@@ -67,36 +66,41 @@ export default class QueryObjectBuilder {
 		return this.queryObject;
 	}
 
-	private transformUnions( conditions: Condition[] ): ( Condition | Condition[] )[] {
-		let currentUnion: Condition[] | null = null;
-		const transformedConditions: ( Condition | Condition[] )[] = [];
-		for ( let i = 0; i < conditions.length; i++ ) {
-			if ( conditions[ i ].conditionRelation === ConditionRelation.Or ) {
-				if ( currentUnion === null ) {
-					const previousCondition = transformedConditions.pop();
-					if ( !previousCondition ) {
-						throw new Error( 'Logic error: empty conditions array when starting union' );
-					}
-					if ( Array.isArray( previousCondition ) ) {
-						throw new Error( 'Logic error: condition array contains union when starting new union' );
-					}
-					currentUnion = [];
-					currentUnion.push( previousCondition );
-				}
-				currentUnion.push( conditions[ i ] );
-			} else {
-				if ( currentUnion !== null ) {
-					transformedConditions.push( currentUnion );
-				}
-				currentUnion = null;
-				transformedConditions.push( conditions[ i ] );
+	private buildConditionTree( conditions: Condition[] ): ( Condition | Condition[] )[] {
+		const rootNode: ( Condition | Condition[] )[] = [];
+		conditions.forEach( ( condition ) => {
+			if ( condition.conditionRelation === 'and' || condition.conditionRelation === null ) {
+				this.attachToRootWithAnd( rootNode, condition );
+				return;
 			}
+			this.attachToPreviousConditionWithOr( rootNode, condition );
+		} );
+
+		return rootNode;
+	}
+
+	private attachToPreviousConditionWithOr( rootNode: ( Condition | Condition[] )[], condition: Condition ): void {
```

Comment:
Nitpick. Can you define "RootNode" type? as ( Condition | Condition[] )[]. This has been repeated a lot and I also like documentation on what root node is and stuff like that too. Of course, this can happen in a follow up.

Reply:
Maybe even as a class having methods like "node.lastElementIsSingleCondition()"

Reply:
> Nitpick. Can you define "RootNode" type? as ( Condition | Condition[] )[]. This has been repeated a lot and I also like documentation on what root node is and stuff like that too. Of course, this can happen in a follow up.

Good point, will do!

> Maybe even as a class having methods like "node.lastElementIsSingleCondition()"

Yeah, I thought about extracting classes here, but it felt like overkill because we are not going to add any more complexity to it in this hike (i.e. no deeper levels of nesting conditions).
And we have no idea if we will ever add those additional nesting levels or when that will happen. So, to me it would seem rather [yagni](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) for now.